File: Examples/HorizonSDKDemo/HorizonSDKDemo/ViewController.m
//
//  ViewController.m
//  HorizonSDKDemo
//
//  Created by Stelios Petrakis on 2/5/15.
//  Copyright (c) 2015 Horizon Video Technologies. All rights reserved.
//

#import "ViewController.h"

@import HorizonSDK;
@import AssetsLibrary;

@interface ViewController () <HVTCameraDelegate>

@property (nonatomic) HVTCamera *camera;

@property (weak, nonatomic) IBOutlet HVTView *preview;
@property (weak, nonatomic) IBOutlet UIButton *recordButton;

@end

@implementation ViewController

#pragma mark - View lifecycle 

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self.recordButton.layer setCornerRadius:33.];
    
    self.camera = [HVTCamera new];
    [self.camera setDelegate:self];
    [self.camera addView:self.preview];
    
    [self authorizeCamera];
}

- (BOOL)prefersStatusBarHidden
{
    return YES;
}

#pragma mark - IB methods

- (IBAction)toggleRecording:(id)sender {
    
    if([self.camera isRecording])
        [self.camera stopRecording];
    else
    {
        NSURL *recordingURL = [[NSURL alloc] initFileURLWithPath:[NSString pathWithComponents:@[NSTemporaryDirectory(), @"Movie.MOV"]]];

        [self.camera startRecordingWithMovieURL:recordingURL];
    }
}

#pragma mark - Private methods

- (void)authorizeCamera
{
    [AVCaptureDevice requestAccessForMediaType:AVMediaTypeVideo
                             completionHandler:^(BOOL granted) {
                                 
                                 dispatch_async(dispatch_get_main_queue(), ^{
                                     [self authorizeMicrophone];
                                 });
                             }];
}

- (void)authorizeMicrophone
{
    [AVCaptureDevice requestAccessForMediaType:AVMediaTypeAudio
                             completionHandler:^(BOOL granted) {
                                 
                                 dispatch_async(dispatch_get_main_queue(), ^{
                                     [self.camera startRunning];
                                 });
                             }];
}

#pragma mark - HVTCameraDelegate

- (void)hvtCameraRecordingDidStart:(HVTCamera *)hvtCamera {
    
    [self.recordButton setTitle:@"Stop" forState:UIControlStateNormal];
}

- (void)hvtCamera:(HVTCamera *)hvtCamera didUpdateRecordingDuration:(NSTimeInterval)duration{
    
    NSUInteger seconds = (NSUInteger)duration;
    
    NSString *newDuration = @"00:00";
    
    if(seconds > 3600.0)
        newDuration = [NSString stringWithFormat:@"%02lu:%02lu:%02lu", seconds/3600UL, ((seconds / 60UL) % 60UL), seconds % 60UL];
    else
        newDuration = [NSString stringWithFormat:@"%02lu:%02lu", seconds / 60UL, seconds % 60UL];
    
    [self.recordButton setTitle:newDuration forState:UIControlStateNormal];
}

- (void)hvtCamera:(HVTCamera *)hvtCamera recordingDidFailWithError:(NSError *)error {
    
    [self.recordButton setTitle:@"Record" forState:UIControlStateNormal];
}

- (void)hvtCamera:(HVTCamera *)hvtCamera didStopRecordingWithMetadata:(NSDictionary*)metadata {
    
    [self.recordButton setTitle:@"Record" forState:UIControlStateNormal];
    
    NSURL *movieURL = [metadata objectForKey:HVTMetadataMovieURLKey];
    
    ALAssetsLibrary *library = [ALAssetsLibrary new];
    
    [library writeVideoAtPathToSavedPhotosAlbum:movieURL
                                completionBlock:^(NSURL *assetURL, NSError *error) {
        
        [[NSFileManager defaultManager] removeItemAtURL:movieURL
                                                  error:NULL];
    }];
}

- (void)hvtCamera:(HVTCamera *)hvtCamera didUpdateParams:(HVTParams)params { }

@end

File: Examples/SwiftHorizonSDKDemo/SwiftHorizonSDKDemo/ViewController.swift
//
//  ViewController.swift
//  HorizonSDKSwiftSample
//
//  Created by Stelios Petrakis on 2/25/15.
//  Copyright (c) 2015 Horizon Video Technologies. All rights reserved.
//

import UIKit
import Photos
import HorizonSDK

class ViewController : UIViewController, HVTCameraDelegate {
    
    var camera : HVTCamera!
    
    @IBOutlet weak var preview: HVTView!
    
    @IBOutlet weak var recordButton: UIButton!
    
    // MARK: View lifecycle
    
    override func viewDidLoad() {
        
        super.viewDidLoad()
        
        recordButton.layer.cornerRadius = 30.0
        
        preview.fillMode = .aspectFill
        
        camera = HVTCamera()
        camera.add(preview)
        camera.delegate = self
        
        AVCaptureDevice.requestAccess(for: AVMediaType.video, completionHandler: { (Bool) -> Void in
            
            AVCaptureDevice.requestAccess(for: AVMediaType.audio, completionHandler: { (Bool) -> Void in
                
                self.camera.startRunning();
            });
        });
    }
    
    override var prefersStatusBarHidden: Bool {
        get {
            return true
        }  
    }
    
    // MARK: IB methods
    
    @IBAction func recordTapped(_ sender: AnyObject) {
        
        if(camera.isRecording()) {
            
            camera.stopRecording()
        }
        else {
            
            let filePath        = NSString.path(withComponents: [NSTemporaryDirectory(), "Movie.mov"])
            let recordingURL    = NSURL(fileURLWithPath: filePath)
            
            camera.startRecording(withMovieURL: recordingURL as URL!)
        }
    }
    
    // MARK: Private methods
    
    func removeTempMovie(movieURL: URL!) {
        do {
            try FileManager.default.removeItem(at: movieURL)
        } catch _ {
        }
    }
    
    // MARK: HVTCameraDelegate methods
    
    func hvtCameraRecordingDidStart(_ hvtCamera: HVTCamera!) {
        
        recordButton.setTitle("Stop", for: .normal)
    }
    
    func hvtCamera(_ hvtCamera: HVTCamera!, didUpdateRecordingDuration duration: TimeInterval) {
        
        let seconds = Int(duration)
        var newDuration = "00:00"
        
        if(seconds > 3600) {
            
            newDuration = NSString(format: "%02lu:%02lu:%02lu", seconds/3600, ((seconds / 60) % 60), seconds % 60) as String
        }
        else {
            
            newDuration = NSString(format: "%02lu:%02lu", seconds / 60, seconds % 60) as String
        }
        
        recordButton.setTitle(newDuration, for: .normal)
    }
    
    func hvtCamera(_ hvtCamera: HVTCamera!, recordingDidFailWithError error: Error!) {
        
        recordButton.setTitle("Record", for: .normal)
    }
    
    func hvtCamera(_ hvtCamera: HVTCamera!, didStopRecordingWithMetadata metadata: [AnyHashable : Any]!) {
        
        recordButton.setTitle("Record", for: .normal)
        
        let metadataDict = metadata as NSDictionary
        let movieURL = metadataDict.object(forKey: HVTMetadataMovieURLKey) as! URL

        PHPhotoLibrary.requestAuthorization({ (status:PHAuthorizationStatus) in
            
            if(status != .authorized) {
                return
            }
          
            PHPhotoLibrary.shared().performChanges({
                PHAssetChangeRequest.creationRequestForAssetFromVideo(atFileURL: movieURL)
            }, completionHandler: { (success, error) in
                
                self.removeTempMovie(movieURL: movieURL)
            })
        })
    }
    
    func hvtCamera(_ hvtCamera: HVTCamera!, didUpdate params: HVTParams) {
        recordButton.transform = CGAffineTransform(rotationAngle: CGFloat(params.angle))
    }
}

